### PR TITLE
Implement the $(LAUNCHER) keyword

### PR DESCRIPTION
**Pull request by koning**
implements #59

### CODE REVIEW

**Comment by ben-bay** [comment]
Wow! Checking out to test right now..

**Comment by ben-bay** [comment]
To run the new `flux_par.yaml`, is there any particular machine to be on?

**Comment by ben-bay** [comment]
Currently getting this error:

```
[2019-12-13 15:31:23,027: INFO] Received task: merlin.common.tasks.merlin_step[9c5e2e08-56e7-4c2f-936f-b6e905cda600]
[2019-12-13 15:31:23,060: ERROR] Chord callback for '8e598bc8-5959-4ea3-b208-6678fda30a22' raised: ChordError("Dependency 9c5e2e08-56e7-4c2f-936f-b6e905cda600 raised KeyError('batch_type')")
Traceback (most recent call last):
  File "/g/g13/bay1/merlin/venv_merlin_py3_7/lib/python3.7/site-packages/celery/app/trace.py", line 385, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/g/g13/bay1/merlin/venv_merlin_py3_7/lib/python3.7/site-packages/celery/app/trace.py", line 648, in __protected_call__
    return self.run(*args, **kwargs)
  File "/g/g13/bay1/merlin/venv_merlin_py3_7/lib/python3.7/site-packages/celery/app/base.py", line 472, in run
    return task._orig_run(*args, **kwargs)
  File "/g/g13/bay1/DELETE/merlin/merlin/common/tasks.py", line 116, in merlin_step
    result = step.execute(config)
  File "/g/g13/bay1/DELETE/merlin/merlin/study/step.py", line 156, in execute
    default_batch_type = adapter_config.pop("batch_type")
KeyError: 'batch_type'
```

**Comment by lucpeterson** [comment]
I think this is a great start, but needs a fair amount of testing

**Comment by ben-bay** [comment]
> Can we think of adding some tests to see that this does what we expect?

Sounds reasonable.

**Comment by koning** [comment]
You need a batch allocation to test properly, but you can test the substitution with a --dry --local run. @ben-bay try it now, I added the batch_type keyword to the other default configs. For the flux_par.yaml you will need two nodes.

File: merlin/study/script_adapter.py [hunk 1]
```diff
@@ -35,13 +35,160 @@
...
+
+        :param **kwargs: A dictionary with default settings for the adapter.
+        """
+        super(MerlinLSFScriptAdapter, self).__init__(**kwargs)
+
+        self._cmd_flags = {
+            "cmd": "srun",
+            "ntasks": "-n",
+            "nodes": "-N",
+            "cores per task": "-c",
```
Comment (ben-bay) [inline merlin/study/script_adapter.py]: What would you think about replacing `cores per task` with `cores_per_task`, for consistency? (`task_queue`, `max_retries`.) Same for `gpus per task`

**Comment by ben-bay** [comment]
Looking more at this now. Question: how is the abstraction of `$(LAUNCHER)` commanded to be flux, lsf, or slurm in a workflow?

File: CHANGELOG.md [hunk 2]
```diff
@@ -28,9 +38,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Development dependencies install via pip: `pip install "merlinwf[dev]"`.
 - `merlin status <yaml spec>` that returns queues, number of connected
-  workers and number of unused tasks in each of those queues.
-- `merlin example` cli command, which allows users to start running the
-  examples immedately (even after pip-installing).
```
Comment (ben-bay) [inline CHANGELOG.md]: Don't think we want to delete these entries, looks like a merge mistake.

Comment (koning) [reply CHANGELOG.md]: OK, I'll look at his.

File: merlin/study/script_adapter.py (hunk 1 above)
Comment (koning) [reply merlin/study/script_adapter.py]: That is from maestro:
https://github.com/LLNL/maestrowf/blob/1bdcf987143e3eabe4235cc913ea910fe16cafca/maestrowf/interfaces/script/slurmscriptadapter.py#L92

File: CHANGELOG.md (hunk 2 above)
Comment (koning) [reply CHANGELOG.md]: Should be fixed now.

File: workflows/flux/flux_local.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
+description:
```
Comment (ben-bay) [inline workflows/flux/flux_local.yaml]: Do we want to add this file to `merlin/examples/workflows/` instead? That's where the other flux workflows currently are.

File: merlin/study/script_adapter.py [hunk 4]
```diff
@@ -35,13 +35,160 @@
...
+            "walltime",
+            "reservation",
+            "restart",
+            "task_queue",
+            "max_retries",
+            "pre",
+            "post",
+            "depends",
+        ]
+        self._unsupported = set(new_unsupported)
```
Comment (ben-bay) [inline merlin/study/script_adapter.py]: Isn't this equivalent to just using curly brackets instead? `new_unsupported = {...`

Comment (koning) [reply merlin/study/script_adapter.py]: That file was added before the move, this PR is from Nov/19.

Comment (koning) [reply merlin/study/script_adapter.py]: The previous comment was for the flux_local.yaml file.

Comment (koning) [reply merlin/study/script_adapter.py]: Yes, it was left over from having two lists , then making a set.

**Pull request merged**